### DISCUSSION TITLE
Loading a 1B row CSV: Stuck with one core at 97% for hours

### DISCUSSION THREAD

**Discussion by ktk** [duckdb/duckdb#10583] [Q&A] [answered]
Hi,

First, fantastic project, thanks for all the work!

I'm doing an extended version of the [1BRC in SQL with DuckDB](https://rmoff.net/2024/01/03/1%EF%B8%8F%E2%83%A3%EF%B8%8F-1brc-in-sql-with-duckdb/) challenge. I extended it a bit to add a dimension table & added timestamps to the observations itself so I can slice & dice on that. Everything works great so far, but I don't fully understand the loading process. I generated a single CSV file with 1B rows and I load it that way:

```
duckdb "$DATABASE_FILE" <<SQL
-- Create the cities table with cityId as PRIMARY KEY
CREATE TABLE cities (
    city VARCHAR,
    cityLabel VARCHAR,
    cityId VARCHAR PRIMARY KEY
--    population INT,
--    coord VARCHAR
);

-- Create the observation table
-- Note: Temp as REAL for decimal temperatures, Timestamp as TIMESTAMP for ISO 8601 datetime strings
CREATE TABLE observation (
    id INTEGER PRIMARY KEY,
    cityId VARCHAR,
    temp REAL, -- REAL is suitable for decimal values like temperatures
    timestamp TIMESTAMP, -- TIMESTAMP for ISO 8601 datetime strings
    FOREIGN KEY (cityId) REFERENCES cities(cityId)
);

-- Load data into cities from CSV
COPY cities FROM '$CSV_DIM' WITH (HEADER TRUE, DELIMITER ';');

-- Load data into observation from CSV
COPY observation FROM '$CSV_OBS' WITH (HEADER TRUE, DELIMITER ';');
SQL

```
I noticed that the larger the CSV file, the longer it takes to do the last few % on the progress bar. Now I'm stuck on a fast server with NVMe SSD at 97% for around 7 hours. I see that only one of the cores is at 100%. I've been reading the docs on loading CSVs, it does mention that I could load multiple files at the same time but I couldn't find any explanation on why that stall might happen with large datasets. Is this the final transaction to write on disk?

Am I doing something wrong and/or could I improve the situation by splitting the file into larger chunks and load them with a list of files instead?

I do want the index on disk, not just in-memory as I use it for something else later.

**Comment by szarnyasg** [comment]
Hi Adrian, which version of DuckDB are you using? The [nightly version](https://duckdb.org/docs/installation/?version=main&environment=cli&installer=binary&platform=macos) has a fully revamped CSV reader.
Gabor

**Reply by ktk** [reply]
Hi Gabor,

I'm on v0.9.2 3c695d7ba9, will try nightly and report back here thanks!

**Reply by ktk** [reply]
ok I'm on it. Interestingly CPU time of the new process (`duckdbN` is latest nightly) ~~shot up way above the old process that runs for hours by now, in just a few minutes~~:

<img width="738" alt="Screenshot 2024-02-12 at 15 16 02" src="https://github.com/duckdb/duckdb/assets/583021/7b3b2671-765a-4672-8d39-0ec936b2295b">

It's also at 97% now. Will let it run for the moment unless you tell me otherwise.

**Reply by ktk** [reply]
Ah and if necessary I can make the whole data available, it's really just synthetic data.

**Reply by ktk** [reply]
eh never mind, once I posted it I see there is a `h` in the old one, sorry about.

**Reply by ktk** [reply]
Update, I stopped both processes last night and changed the approach. I now feed 5M CSV lines per chunk with this loop:

```
# Find all chunk_xx files in the data directory and load them
find data -name 'chunk_*' | sort | while read chunk_file; do
    echo "Loading $chunk_file into observation table..."
    duckdbN "$DATABASE_FILE" <<SQL
COPY observation FROM '$chunk_file' WITH (HEADER FALSE, DELIMITER ';');
SQL
done
```

This seems to work fine initially but I think it's getting slower and slower with each chunk. I'm now on chunk 30 so 150M lines in total and I don't think this will finish anytime soon. This is after roughly 15h, with the nightly version.

What are other experiences in limits of DuckDB right now? In memory does not seem to be the problem according to the link I posted initially. But in my use-case I *need* a duckdb in file as I use it to wrap another query language on top. I'm doing an extended version of what I posted here: https://www.linkedin.com/pulse/rdf-olap-cubes-using-duckdb-ontop-adrian-gschwend-9qy0e/

Should I make the files available? It's 45GB uncompressed, I think around 15 bzipped. I could put it on a S3.

**Comment by FrancoisLepoutre** [comment, marked as answer]
Hi,

A dummy remark about building indexes. My experience has been that when building database index-es for decision support it is sometimes more effective to build the indexing material AFTER the initial data feed.

The Duckdb documentation points to something similar:

> Best Practices:
> Only use primary keys, foreign keys, or unique constraints, if these are necessary for enforcing constraints on your data.
> Do not define explicit indexes unless you have highly selective queries.
> **If you define an ART index, do so after bulk loading the data to the table.**

https://duckdb.org/docs/guides/performance/indexing

Not tested by myself. But I'd glad to hear whether this sensible generic piece of advice for OLAP-centric db engines applies to your use-case.

**Reply by ktk** [reply]
Interesting thought, thanks! I did read that part as well yesterday and I was thinking if I should drop them all together. I don't really have a reason for it anymore, I added it because I noticed some data problems in the synthetic input I generated. But now that it's solved in the generator, there is no real need anymore.

Will try without keys & report here if it influences the DB generation!

**Reply by ktk** [reply]
Good news, that was the issue! It loads super fast without primary/foreign keys!

It says in the docs they could be added later but I don't see an according `ALTER TABLE` support for modifying that, how would I do that?

**Reply by ktk** [reply]
Ah and as one remark: I think it would be good to add the fact that keys can completely destroy loading time for large datasets to the documentation on https://duckdb.org/docs/guides/performance/indexing

**Reply by szarnyasg** [reply]
@ktk `ALTER TABLE` is not yet supported:
```sql
D CREATE TABLE customers (id INT, name VARCHAR);
Run Time (s): real 0.001 user 0.000615 sys 0.000559
D ALTER TABLE customers ADD PRIMARY KEY (id);
Run Time (s): real 0.000 user 0.000164 sys 0.000016
Error: Not implemented Error: ALTER TABLE option not supported yet!
```

> Ah and as one remark: I think it would be good to add the fact that keys can completely destroy loading time for large datasets to the documentation on https://duckdb.org/docs/guides/performance/indexing

Good point. It's documented on the [schema page](https://duckdb.org/docs/guides/performance/schema#microbenchmark-the-effect-of-primary-keys) but it should be cross-linked in the indexing page.

**Reply by szarnyasg** [reply]
https://github.com/duckdb/duckdb-web/pull/2131

**Comment by FrancoisLepoutre** [comment]
Hi again Adrian,
Thanks for feeding back:-)

**Reply by ktk** [reply]
FYI that's what I was building: https://www.linkedin.com/pulse/scaling-sparql-querying-billion-observations-ontop-duckdb-gschwend-myghf
